addons/account/tests/test_account_inalterable_hash.py

```diff
@@ -188,3 +188,81 @@ def test_account_move_hash_versioning_v1_to_v2(self):
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-03", amounts=[1000, 2000])
+        )
+        moves_v2.with_context(hash_version=2).action_post()
+
+        # invalidate cache
+        moves_v2[0].line_ids[0].invalidate_recordset()
+
+        integrity_check = moves_v2.company_id._check_hash_integrity()['results'][0]
+        self.assertRegex(integrity_check['msg_cover'], f'Corrupted data on journal entry with id {moves_v2[0].id}')
+
+    def test_account_move_hash_versioning_3(self):
+        """
+        Test that version 3 covers correctly the case that has floating-point representation issues
+        """
+        self.init_invoice("out_invoice", self.partner_a, "2023-01-01", amounts=[1000, 2000],
+                          post=True)  # Not hashed
+        self.company_data['default_journal_sale'].restrict_mode_hash_table = True
+        moves_v3 = (
+            self.init_invoice("out_invoice", self.partner_a, "2023-01-01", amounts=[30*0.17, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-02", amounts=[1000, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-03", amounts=[1000, 2000])
+        )
+        moves_v3.action_post()
+
+        # invalidate cache
+        moves_v3[0].line_ids[0].invalidate_recordset()
+
+        integrity_check_v3 = moves_v3.company_id._check_hash_integrity()['results'][0]
+        self.assertRegex(integrity_check_v3['msg_cover'], f'Entries are hashed from {moves_v3[0].name}.*')
+
+    def test_account_move_hash_versioning_v2_to_v3(self):
+        """
+        We are updating the hash algorithm. We want to make sure that we do not break the integrity report.
+        This test focuses on the case with version 2 and version 3.
+        """
+        self.init_invoice("out_invoice", self.partner_a, "2023-01-01", amounts=[1000, 2000],
+                          post=True)  # Not hashed
+        self.company_data['default_journal_sale'].restrict_mode_hash_table = True
+        moves_v2 = (
+            self.init_invoice("out_invoice", self.partner_a, "2023-01-01", amounts=[1000, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-02", amounts=[1000, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-03", amounts=[1000, 2000])
+        )
+        moves_v2.with_context(hash_version=2).action_post()
+        fields_v2 = moves_v2._get_integrity_hash_fields()
+
+        moves_v3 = (
+            self.init_invoice("out_invoice", self.partner_a, "2023-01-01", amounts=[1000, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-02", amounts=[1000, 2000])
+            | self.init_invoice("out_invoice", self.partner_b, "2023-01-03", amounts=[1000, 2000])
+        )
+        moves_v3.with_context(hash_version=3).action_post()
+        fields_v3 = moves_v3._get_integrity_hash_fields()
+        self.assertEqual(fields_v2, fields_v3)
```

Comment:
Not sure why we need to check this.
Note that `moves_v2` and `moves_v3` have the same context (without `hash_version` set, therefore there is no reason for `_get_integrity_hash_fields` to return something different-

Reply:
> Not sure why we need to check this.

For the same reason there was a test v1 to v2, i.e. check that the integrity report goes through all possible versions without errors

> Note that moves_v2 and moves_v3 have the same context (without hash_version set, therefore there is no reason for _get_integrity_hash_fields to return something different-

Indeed, perhaps @yosa-odoo could test `_getattrstring` instead

Reply:
> For the same reason there was a test v1 to v2

I mean the specific assert I'm commenting on. We don't really care about the return result of private functions, we care about user flows.

> Indeed, perhaps @yosa-odoo could test `_getattrstring` instead

? :suspect:

Reply:
Not sure to understand the `_getattrstring` but we can indeed only check that the report is flawless with use of version 2 and version 3